### ISSUE TITLE
Login via Nextcloud fails with SSL or CSRF error - Internal Server Error

### ISSUE THREAD

**Issue by yoitcc** [loomio/loomio#7432]
Hello :)

I am using the docker image from loomio/loomio-deploy (updated today to 2.7.1), I am very happy about using it, it works pretty seamlessly. What I didn't get to work (today and a few times before) is logging in via my own Nextcloud (also dockerized, version 21.0.0, SSL/HTTPS setup is working (is valid and uses Nextcloud recommended headers). 

I tried activating the Nextcloud-Login a few times long before, never worked or showed me a reasonable path to solve it (IRC it didn't open my nextcloud page or didn't accept the login). 

Today (in the progress of updating) I wanted to try it again, I added loomio as an OAuth Client in Nextcloud and set the three enviroment variables (NEXTCLOUD_HOST / NEXTCLOUD_APP_KEY / NEXTCLOUD_APP_SECRET)  in the file 'env'. After a docker down and up (and a private session in Firefox) it showed me the Button for the Nextcloud Login, forwarded me to the Login Page (Login worked), forwarded me to the "Grant Access" Page of Nextcloud and then sent me back to loomio, presenting me with a "Internal Server Error" (and not being logged in after a refresh).

This was farther than I ever got (and presumably puts the problem within loomio or loomio-deploy) so I checked the Loomio logs, and finally it showed me something search- and pasteable ;)
This lines highlight the problem (as far as I can tell):
> loomio-gang-channels | have current user! null []
> loomio-gang-app | source=rack-timeout id=cabf9d91-62f7-43a8-aa13-becb12ebee0f timeout=15000ms state=ready
> loomio-gang-app | method=GET path=/nextcloud/oauth format=html controller=Identities::NextcloudController action=oauth status=302 duration=17.96 view=0.00 db=0.70 location=https://nextcloud.myowndomain.com/index.php/apps/oauth2/authorize
> loomio-gang-app | source=rack-timeout id=cabf9d91-62f7-43a8-aa13-becb12ebee0f timeout=15000ms service=22ms state=completed
> loomio-gang-channels | have current user! null []
> loomio-gang-channels | catchup: { data: {}, callback: [Function (anonymous)] }
> loomio-gang-app | source=rack-timeout id=8c153a71-c4a8-45e9-bad1-94ceed808b29 timeout=15000ms state=ready
> loomio-gang-app | opening connection to nextcloud.myowndomain.com:443...
> loomio-gang-app | opened
> loomio-gang-app | starting SSL for nextcloud.myowndomain.com:443...
> loomio-gang-app | Conn close because of connect error SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate)
> loomio-gang-app | method=GET path=/nextcloud/authorize format=html controller=Identities::NextcloudController action=create status=500 error='OpenSSL::SSL::SSLError: SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate)' duration=41.59 view=0.00 db=0.00
> loomio-gang-app |   
> loomio-gang-app | **OpenSSL::SSL::SSLError (SSL_connect returned=1 errno=0 state=error: certificate verify failed (unable to get local issuer certificate)):**
> loomio-gang-app |   
> loomio-gang-app | app/extras/clients/request.rb:17:in `response'
> loomio-gang-app | app/extras/clients/request.rb:12:in `perform!'
> loomio-gang-app | app/extras/clients/base.rb:59:in `block in perform'
> loomio-gang-app | app/extras/clients/base.rb:59:in `tap'
> loomio-gang-app | app/extras/clients/base.rb:59:in `perform'
> loomio-gang-app | app/extras/clients/base.rb:22:in `post'
> loomio-gang-app | app/extras/clients/nextcloud.rb:4:in `fetch_access_token'
> loomio-gang-app | app/controllers/identities/base_controller.rb:62:in `identity_params'
> loomio-gang-app | app/controllers/identities/base_controller.rb:36:in `identity'
> loomio-gang-app | app/controllers/identities/base_controller.rb:8:in `create'
> loomio-gang-app | app/helpers/locales_helper.rb:8:in `use_preferred_locale'
> loomio-gang-app | app/helpers/locales_helper.rb:3:in `process_time_zone'
> loomio-gang-app | source=rack-timeout id=8c153a71-c4a8-45e9-bad1-94ceed808b29 timeout=15000ms service=49ms state=completed
> loomio-gang-app | source=rack-timeout id=5a82979c-da99-4f36-8848-084958c8c0f0 timeout=15000ms state=ready
> loomio-gang-app | Can't verify CSRF token authenticity.
> loomio-gang-app | method=POST path=/bhoy/events format=*/* controller=Ahoy::EventsController action=create status=200 duration=96.89 view=0.23 db=90.61
> loomio-gang-app | source=rack-timeout id=5a82979c-da99-4f36-8848-084958c8c0f0 timeout=15000ms service=99ms state=completed

I also searched for this message, but it doesn't seem connected to the login process: 
> loomio-gang-app | Can't verify CSRF token authenticity.

--> that's what I found: https://stackoverflow.com/questions/35181340/rails-cant-verify-csrf-token-authenticity-when-making-a-post-request

Full log (starting with opening the page loomio.myowndomain.com)
[loomio-nextcloud-loginfail.txt](https://github.com/loomio/loomio/files/6177697/loomio-nextcloud-loginfail.txt)

The search for the OpenSSL-line mainly turns up posts talking about missing root certificates (on the system/in the docker image), so I went in the container loomio-app to check:
> root@d95b23428b17a:/loomio# ls /etc/ssl/certs/ -la | grep -i isrg
> lrwxrwxrwx 1 root root     16 Feb  9 04:35 4042bcee.0 -> ISRG_Root_X1.pem
> lrwxrwxrwx 1 root root     51 Feb  9 04:35 ISRG_Root_X1.pem -> /usr/share/ca-certificates/mozilla/ISRG_Root_X1.crt

Firefox shows me the  "ISRG Root X1" Certificate as root for the LetsEncrypt certificate from Traefik (v2.4) for the domain nextcloud.myowndomain.com (and loomio.myowndomain.com).

On the other hand, inside the loomio-app container (after reading this again: https://stackoverflow.com/questions/36966650/ruby-nethttp-responds-with-opensslsslsslerror-certificate-verify-failed);
> root@d95b80e8b17a:/loomio# curl https://nextcloud.myowndomain.com
> curl: (60) SSL certificate problem: unable to get local issuer certificate

The error stays the same after update-ca-certificates...
> root@d95b80e8b17a:/loomio# update-ca-certificates 
> Updating certificates in /etc/ssl/certs...
> 0 added, 0 removed; done.
> Running hooks in /etc/ca-certificates/update.d...
> done.

Another result of my search was (https://stackoverflow.com/questions/9199660/why-is-ruby-unable-to-verify-an-ssl-certificate), talking about the ruby approach (only using certificates that are specifically needed), but the Google, Facebook, Slack,... integrations of loomio make me think that is not really the problem here.

And thats where I am stuck, there are 2 closed issues in loomio-deploy that turn up when searching for nextcloud, but they are rather unrelated. Besides that I did only find the merge #6436 (https://github.com/loomio/loomio/pull/6436), which added support for the nextcloud login.

So to sum it up:
- I expected the Login via Nextcloud to work as anticipated
- Nextcloud verified and granted access to loomio 
- Loomio throws an 'Internal Server Error', with "certificate verify failed (unable to get local issuer certificate" in the logs

Do you please have any advice on how to get it working? Can I help with any more information?
Thank you for your patient reading ;)
Best Wishes,
Oswald

**Comment by robguthrie** [comment]
Hi Oswald, thanks for the detailed write up. And sorry for the slow reply.

I've seen the same error related to sending emails, and the solution is to tell OpenSSL to turn off certificate validation.

The error is coming from the http request, which uses HTTParty to make requests.

I found this page:
https://stackoverflow.com/questions/20955177/how-do-i-make-httparty-ignore-ssl

So I've added two ENV's which you can use to disable SSL cert verification in HTTParty, which nextcloud uses.

```
Clients::Request = Struct.new(:method, :url, :params) do
  include HTTParty
  default_options.update(verify: false) if ENV['SSL_VERIFY_FALSE']
  default_options.update(verify_peer: false) if ENV['SSL_VERIFY_PEER_FALSE']
```

to master, which means you'll need to use the 'latest' docker image. (edit your `.env` file and change LOOMIO_CONTAINER_TAG to latest). (then docker-compose pull)

And to use them, edit your `env` file (not to be confused with the .env file) and add SSL_VERIFY_FALSE=1 or SSL_VERIFY_PEER_FALSE=1 or both, then docker-compose down, and docker-compose up

I hope this all helps!

**Comment by yoitcc** [comment]
Hello Robert!

I am absolutely amazed by your reply (and that you changed code so fast) and I see it as really fast! Thank you very much for your time and effort :)

Sadly, it didn't solve the problem, but it did change the error ;)
Now in the logs there is a HTML-formatted error about the CSRF-Checksum failing (browser still says "internal server error"; cookies are enabled, the second time nextcloud still had me logged in).
The first log is just with SSL_VERIFY_FALSE=1, the second is additionally with SSL_VERIFY_PEER_FALSE=1:
[loomio-issue-ssl-verify-false-1.txt](https://github.com/loomio/loomio/files/6204924/loomio-issue-ssl-verify-false-1.txt)
[loomio-issue-ssl-verify-false-1+ssl-verify-peer-false-1.txt](https://github.com/loomio/loomio/files/6204926/loomio-issue-ssl-verify-false-1%2Bssl-verify-peer-false-1.txt)

Can I help you with any more tests? Should I check if setting SSL_VERIFY_FALSE=0 and SSL_VERIFY_PEER_FALSE=1 (or vice-versa) changes anything?

@AlbanBedel You merged the feature into loomio, do you have the nextcloud login working on your side? Are you still interested in working on this topic? 

Thank you all very much for your time and efforts! Be sure that it is appreciated in Vienna ;)
Have a nice day!
Oswald

**Comment by AlbanBedel** [comment]
According to your last log loomio sent a request to your NC server, but an OPNSense server answered the request. So it sound like the request is blocked by your firewall, which might also explain the SSL errors you were seeing.

OAauth2 rely on exchanging URLs between the browser and 2 web servers, for it to work correctly you must make sure that the public URL of your services work correctly from within your containers. This can be tricky when the setup involve docker and/or some NAT forwarding. To debug you can run some test requests using curl from within each container and compare to what you get from a test client.

**Comment by yoitcc** [comment]
Hello again, sry for the long delay, thank you Alban for looking into it, took me a while to get around testing ^^
You are right about the OPNSense Firewall and the NAT-hell of docker. The setup is built as follows:
- A dedicated server running different virtual machines with Proxmox
- One of the virtual machines is the OPNSense Firewall, essentially being the face of the one public IP that my domains (nextcloud and loomio) resolve to
- A second virtual machine hosts the docker services, with ports 80/443 forwarded from the firewall to the dockerized Traefik reverse proxy, handling SSL and "domain-to-container" stuff

There is some problem with name-resolution within the containers, also OPNSense is showing a site warning about DNS Rebinding, I am still in the process of sorting this out, this is just an update on the progress.

Best Regards,
Oswald

**Comment by AlbanBedel** [comment]
You generally want to whitelist your own domains from the DNS rebinding protection as they have a private address when used from inside.

**Comment by robguthrie** [comment]
I don't know what's going on here, but I'm closing the issue because it's old. If there is value in keeping it open, please let me know.

**Comment by yoitcc** [comment]
Yeah, so I don't even remember if I got it working in the end or if we switched to local accounts :see_no_evil: - it certainly is no longer relevant to me :stuck_out_tongue_winking_eye:  

Thank you both for Guidance & Replying!
Have a good one!
